src/messages.rs

```diff
@@ -493,9 +498,18 @@ where
         }
     };
 
-    let mut buf = vec![0u8; len as usize - 4];
+    buffer.put_u8(code);
+    buffer.put_i32(len);
 
-    match stream.read_exact(&mut buf).await {
+    buffer.resize(buffer.len() + len as usize - 4, b'0');
+
+    match stream
+        .read_exact(
+            &mut buffer[starting_point + mem::size_of::<u8>() + mem::size_of::<i32>()
```

Comment:
You're sometimes hardcoding the length of the integer (4 bytes) and sometimes using these functions. Might be best to stick to one or the other.